agent/xds/golden_test.go

```diff
@@ -45,14 +53,38 @@ func golden(t *testing.T, name, subname, got string) string {
 	}
 
 	golden := filepath.Join("testdata", name+suffix)
+
+	// To trim down PRs, we only create per-version golden files if they differ
+	// from the latest version.
+	newestExpected := ""
+	if newestSubname != "" && subname != newestSubname {
+		newestGolden := filepath.Join("testdata", fmt.Sprintf("%s.%s.golden", name, newestSubname))
+		expected, err := ioutil.ReadFile(newestGolden)
+		require.NoError(t, err)
+
+		if string(expected) == got {
+			if *update && got != "" {
```

Comment:
I couldn't figure out what was going on here; the original optional writes the file, then returns the contents, which seems straightforward, and provides a path to regenerate the golden files. But I'm not sure why we delete the golden file, but don't do anything with the newest golden, or why this is guarded by the match with expected.

Perhaps a comment explaining the new workflow, in particular if the version under test isn't the newest.

Reply:
Yeah I'll add a bunch of comments.

The gist is that when you run the tests with the `-update` flag they will erase any golden test files for prior versions that are identical to the golden test file for the latest version of envoy.

During comparison mode, we check for a specific version file if one exists and use that for comparison, but if that does not exist we use the one from the latest version of envoy instead.

Reply:
@markan check the latest commit in the PR for the restructured function that should hopefully be way clearer.